hooks-howto.md

```diff
@@ -0,0 +1,72 @@
+# JavaScript Hooks in the CSI Mock Driver
+The CSI mock driver can call JavaScript hooks during the execution of the CSI
+method calls. These hooks can be configured in an external YAML file that can be
+passed to the mock driver sidecar container as a `ConfigMap` volume.
+
+## Example
+Here's an usage example of the feature. First, prepare a YAML file defining
+some JavaScript hooks. Let's call it `hooks.yaml`:
+
+```
+Globals: |
+  count = 0;
+  console.log("Globals loaded, count set to " + count);
+CreateVolumeStart: |
+  count = count + 1;
+  console.log("CreateVolumeStart: The value of count is " + count);
+  if (count > 2) { __grpcCode.OK; } else {  __grpcCode.DEADLINEEXCEEDED; };
+CreateVolumeEnd: |
+  console.log("CreateVolumeEnd");
+```
+
+Create a `ConfigMap` from the YAML:
+
+```
+kubectl create configmap hooks-config --from-file hooks.yaml
+```
+Configure the CSI mock driver to use it -- these are the relevant parts of the
+sidecar definition YAMLs:
+
+```
+...
+containers:
+  ...
+  - name: csi-mock-plugin
+    ...
+    args:
+      - --hooks-file=/var/lib/hooks/hooks.yaml
+    ...
+    volumeMounts:
+      ...
+      - name: hooks-volume
+        mountPath: /var/lib/hooks
+...
+volumes:
+  - name: hooks-volume
+    configMap:
+      name: hooks-config
+...
+```
+
+Now the driver finds the hooks definitions file and creates a JavaScript VM.
+Just one instance of the VM exists through the life of the driver. It executes
+the pieces of code as per the YAML file. One special case is the `Globals` hook
+that is executed right after loading the file and it's intended for variable
+initialization, common function definitions, etc. There are also pre-defined
+gRPC constants in the `__grpcCode` variable.
+
+Each code snippet is executed at given place in the code: in this example
+`CreateVolumeStart` at the beginning of the `CreateVolume` method in the driver.
+The result of the hook snipped is evaluated: it would be the value of the last
+expression executed. If the result is an integer value it gets interpreted as
+gRPC code and in case it's non-zero (not `code.OK`), the CSI driver method would
+return this as its response error.
+
+In the example: The variable `count` was initialized to zero in the `Globals`
+hook. Each time the `CreateVolume` method is called the driver the hook is
```

Comment:
"Each time the `CreateVolume` method is called the hook is executed"

Reply:
Fixed.